### PR TITLE
multi: implement BIP 341 and 342 a.k.a complete taproot and tapscript consensus verification logic 

### PR DESCRIPTION
**Pull request by Roasbeef**
In this PR, we implement full validation logic for BIP 341 and BIP 342. Along the way, we've created a series of utility functions to be used in wallet libraries to handle things like parsing the control block, generating merkle tapscript inclusion proofs, and computing the two sighash variants introduced in the BIPs. 

See each commit message for a detailed description w.r.t the scope of the incremental changes, as well as notes for reviewers w.r.t improvements to the logic/routines or things to doubly verify. 

This PR builds on a series of PRs including:
  * https://github.com/btcsuite/btcd/pull/1700
  * https://github.com/btcsuite/btcd/pull/1773
  * https://github.com/btcsuite/btcd/pull/1777
  * https://github.com/btcsuite/btcd/pull/1785
  * https://github.com/btcsuite/btcd/pull/1768

## Steps to Graduate from a Draft PR
- [x] Merge in the BIP 340 implementation linked above, along with the new `v2` module for `btcec`. 
- [x] Merge in the modified BIP 9 implementation linked above, that implements the new min activation height vector. 
- [x] Extract the generated test vectors from the `bitcoind` implementation (the BIPs contain no test vectors, but they have some good coverage saved from their fuzz testing) to ensure the logic implemented is compliant with consensus rules. 
- [x] Add additional test coverage for the utillity methods included
- [x] Re-examine the current routine for generating the script tree, to examine more efficient variants, or variants that produce a non-leaning binary tree structure 

 

Fixes everything in https://github.com/btcsuite/btcd/issues/1735, other than the `btcwallet` specific components.

### CODE REVIEW

File: txscript/script.go [hunk 1]
```diff
@@ -37,13 +38,36 @@ const (
 	sigHashMask = 0x1f
 )
 
+const (
+	// TaprootAnnexTag is the tag for an annex. This value is used to
+	// identify the annex during tapscript spends. If there're at least two
+	// elements in the taproot witness stack, and the first byte of the
+	// last element matches this tag, then we'll extract this as a distinct
+	// item.
+	TaprootAnnexTag = 0x05
```
Comment (guggero) [inline txscript/script.go:48]: Isn't this `0x50`? https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#Rationale

File: txscript/script.go [hunk 2]
```diff
@@ -303,11 +327,13 @@ func removeOpcodeByData(script []byte, dataToRemove []byte) []byte {
 // being spent, in addition to the final transaction fee. In the case the
 // wallet if fed an invalid input amount, the real sighash will differ causing
 // the produced signature to be invalid.
-func calcWitnessSignatureHashRaw(scriptSig []byte, sigHashes *TxSigHashes,
+func calcWitnessSignatureHashRaw(subScript []byte, sigHashes *TxSigHashes,
```
Comment (guggero) [inline txscript/script.go:341]: Nice to fix this naming confusion! I always tripped over this, since we set the pub key hash here in case of P2WPKH...

File: txscript/script.go [hunk 3]
```diff
@@ -420,6 +446,216 @@ func CalcWitnessSigHash(script []byte, sigHashes *TxSigHashes, hType SigHashType
...
+	sigHashOpts ...TaprootSigHashOption) ([]byte, error) {
+
+	opts := defaultTaprootSighashOptions()
+	for _, sigHashOpt := range sigHashOpts {
+		sigHashOpt(opts)
+	}
+
+	// If a valid sighash type isn't passed in, then we'll exit early.
+	if !isValidTaprootSigHash(hType) {
+		// TODO(roasbeef): use actual errr here
```
Comment (guggero) [inline txscript/script.go:564]: Is this TODO still relevant? Since there is no `err` returned by `isValidTaprootSigHash`?

File: txscript/engine.go [hunk 4]
```diff
@@ -488,19 +519,82 @@ func (vm *Engine) verifyWitnessProgram(witness [][]byte) error {
...
+			// keyspend validation.
+			rawSig := witness[0]
+			err := VerifyTaprootKeySpend(
+				vm.witnessProgram, rawSig, &vm.tx, vm.txIdx,
+				vm.prevOutFetcher, vm.hashCache, vm.sigCache,
+			)
+			if err != nil {
+				// TODO(roasbeef): proper error
+				return err
+			}
```
Comment (guggero) [inline txscript/engine.go:638]: Looks like this commit doesn't compile because of a missing closing brace here?

File: txscript/taproot.go [hunk 5]
```diff
@@ -1,11 +1,56 @@
...
 	"fmt"
 
+	secp "github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
 	"github.com/btcsuite/btcd/wire"
 )
 
+// LeafVersionMask is the mask applied to the first byte of the control block
+// to extract the committed leaf versions.
+const LeafVersionMask = 0xfe
```
Comment (guggero) [inline txscript/taproot.go:19]: This is already defined as `TaprootLeafMask` in `script.go`.

**Review by guggero — COMMENTED** [review]
Very excited about this PR, amazing work! Did a first quick pass, mostly to get a sense of the size of the diff. I have to say, it's not as large as I expected.
It'll probably grow quite a bit once tests with all test vectors are added.

**Comment by gcsfred2** [comment]
In script.go, calls to `checkScriptParses` have `scriptVersion` hardcoded to 0. Is that ok?

File: txscript/taproot.go [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+	// This simulates revealing a leaf from the largest possible tapscript
+	// tree.
+	ControlBlockMaxSize = ControlBlockBaseSize + (ControlBlockNodeSize *
+		ControlBlockMaxNodeCount)
+)
+
+var (
+	// TaprootTagSigHash is the message tag prefix used to compute the
+	// sighash used to validate taproot schnorr signatures.
+	TaprootTagSigHash = []byte("TapSighhash")
```
Comment (gcsfred2) [inline txscript/taproot.go:59]: @Roasbeef Consider enforcing ASCII encoding to match TaggedHash in Bitcoin's hash.cpp. Bitcoin's hash.cpp assumes one byte per character. UTF-8 encoding does lead to different hashes and to invalid signatures. An alternative is 
`	TaprootTagSigHash = []byte{0x54, 0x61, 0x70, 0x53, 0x69, 0x67, 0x68, 0x61, 0x73, 0x68} // "TapSighhash"
`

**Review by positiveblue — COMMENTED** [review]
Really excited to see btcd supporting taproot! First pass done with a couple of small nits

**Review by wpaulino — COMMENTED** [review]
Great work putting all of this together! I've completed an initial pass in which I mostly focused on spec compat.

The PR will become a bit unwieldy once tests have been added, I think it would be useful to break it up into smaller parts to help reviewers focus on some of the independent changes being added. An example breakdown could be (each point represents its own PR):
* Prep work for Taproot integration
* Taproot key spend path
* Control block construction/verification
* Taproot script spend path (including sig opcode changes)

**Review by guggero — COMMENTED** [review]
Started some end-to-end integrations (see https://github.com/guggero/btcwallet/tree/schnorr-taproot and https://github.com/guggero/lnd/tree/schnorr-taproot-signet) and found a few small-ish issues. Going to do a more in-depth review of this PR again tomorrow, just leaving those comments here.

File: txscript/taproot.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+// of a signature that has already been seen.
+//
+// NOTE: The TxSigHashes MUST be passed in an fully populated.
+func VerifyTaprootKeySpend(witnessProgram []byte, rawSig []byte, tx *wire.MsgTx,
+	inputIndex int, prevOuts PrevOutputFetcher, hashCache *TxSigHashes,
+	sigCache *SigCache) error {
+
+	// First, we'll need to extract the public key from the witness
+	// program.
+	rawKey := extractWitnessV1KeyBytes(witnessProgram)
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/taproot.go:88]: We likely want to tighten up the typing here, to make it clearer if we're expecting the full pk script, or just the raw public key bytes.

File: txscript/taproot.go [hunk 8]
```diff
@@ -261,6 +264,53 @@ func ComputeTaprootOutputKey(internalKey *btcec.PublicKey,
...
 }
 
+// ComputeTaprootKeyNoScript calculates the top-level taproot output key given
+// an internal key, and a desire that the only way an output can be spent is
+// with the keyspend path. This is useful for normal wallet operations that
+// don't need any other additional spending conditions.
+func ComputeTaprootKeyNoScript(internalKey *btcec.PublicKey) *btcec.PublicKey {
+	// We'll compute a custom tap tweak hash that just commits to the key,
+	// rather than an actual root hash.
+	fakeScriptroot := schnorr.SerializePubKey(internalKey)
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/taproot.go:275]: Ah makes sense, we end up effectively including the internal key twice in the hash digest here.

File: txscript/hashcache.go [hunk 9]
```diff
@@ -5,30 +5,273 @@
...
-		HashPrevOuts: calcHashPrevOuts(tx),
-		HashSequence: calcHashSequence(tx),
-		HashOutputs:  calcHashOutputs(tx),
+func NewTxSigHashes(tx *wire.MsgTx,
+	inputFetcher PrevOutputFetcher) *TxSigHashes {
+
+	var sigHashes TxSigHashes
+
+	// Base segwit (witness version v0), and taproot (witness version v1)
+	// differ in how the set of ore-computed cached sighash midstate is
```
Comment (alexbosworth) [inline txscript/hashcache.go:207]:
```suggestion
	// differ in how the set of pre-computed cached sighash midstate is
```

File: txscript/standard.go [hunk 10]
```diff
@@ -381,10 +383,26 @@ func extractWitnessScriptHash(script []byte) []byte {
 	return nil
 }
 
+// extractWitnessV1ScriptHash extracts the witness script hash from the passed
+// script if it is standard pay-to-witness-script-hash script. It will return
+// nil otherwise.
+func extractWitnessV1ScriptHash(script []byte) []byte {
+	// A pay-to-witness-script-hash script is of the form:
+	//   OP_1 OP_DATA_32 <32-byte-hash>
+	if len(script) == 34 &&
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/standard.go:395]: Updated to use those constants. Opted to keep the relationship as is, since it matches the existing instances.

File: txscript/script.go [hunk 11]
```diff
@@ -420,6 +446,216 @@ func CalcWitnessSigHash(script []byte, sigHashes *TxSigHashes, hType SigHashType
...
+		sigMsg.Write(sigHashes.HashPrevOutsV1[:])
+		sigMsg.Write(sigHashes.HashInputAmountsV1[:])
+		sigMsg.Write(sigHashes.HashInputScriptsV1[:])
+		sigMsg.Write(sigHashes.HashSequenceV1[:])
+	}
+
+	// If this is sighash all, or its taproot alias (sig hash default),
+	// then we'll also include the pre-computed digest of all the outputs
+	// of the transaction.
+	if hType == SigHashDefault || hType == SigHashAll {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/script.go:610]: IMO the BIP text here is confusing as it states the sighash types by process of elimination rather than being more explicit, but updated to match more directly. It directs implementation to use a bitmask when it isn't actually necessary as you can just compare the raw types, unlike sighash anyone. 

Nevertheless, changed to match the BIP.

File: txscript/sign.go [hunk 12]
```diff
@@ -64,6 +66,57 @@ func WitnessSignature(tx *wire.MsgTx, sigHashes *TxSigHashes, idx int, amt int64
...
+
+	// Finally, append the sighash type to the final sig.
+	return append(signature.Serialize(), byte(hashType)), nil
+}
+
+// TaprootWitnessSignature returns a valid witness stack that can be used to
+// spend the key-spend path of a taproot input as specified in BIP 342.
+//
+// TODO(roasbeef): add support for annex even tho it's non-standard?
+func TaprootWitnessSignature(tx *wire.MsgTx, sigHashes *TxSigHashes, idx int,
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/sign.go:106]: Good point, modified to return the proper signature encoding, and also modified the godoc. This way there's less burden on the caller to do the right thing if they want the 64-byte signature.

File: txscript/script.go [hunk 13]
```diff
@@ -462,12 +467,62 @@ const (
...
 // taprootSigHashOptions houses a set of functional options that may optionally
 // modify how the taproot/script sighash digest algorithm is implemented.
 type taprootSigHashOptions struct {
+	// extFlag denotes the current message digest extension being used. For
+	// top-level script spends use a value of zero, while each tapscript
+	// version can define its own values as well.
+	extFlag sigHashExtFlag
+
+	// tapLeafHash is the hash of the tapscript leaf as defined in BIP 341.
+	// This should be h_tapleaf(compactSizeOf(script) || script).
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/script.go:487]: Fixed, FWIW was just the comment, we compute it properly in `engine.go` when we detect a tapscript spend.

File: txscript/sigvalidate.go [hunk 14]
```diff
@@ -24,6 +24,197 @@ type signatureVerifier interface {
...
+
+	subScript []byte
+
+	hashType SigHashType
+}
+
+// parseBaseSigAndPubkey attempts to parse a signature and public key according
+// to the base consensus rules, which expect an 33-byte public key and DER or
+// BER encoded signature.
+func parseBaseSigAndPubkey(strictEncoding bool, pkBytes, fullSigBytes []byte,
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/sigvalidate.go:50]: True, moved the flag checks into this function.

File: txscript/opcode.go [hunk 15]
```diff
@@ -1972,90 +1972,37 @@ func opcodeCheckSig(op *opcode, data []byte, vm *Engine) error {
...
-	} else {
-		valid = signature.Verify(hash, pubKey)
-	}
+	// TODO(roasbeef): verify NULLFAIL semantics as relates to constructors
+	// above and empty sig vectors
+	valid := sigVerifier.Verify()
 
-	if !valid && vm.hasFlag(ScriptVerifyNullFail) && len(sigBytes) > 0 {
+	switch {
+	case !valid && vm.hasFlag(ScriptVerifyNullFail) && len(fullSigBytes[1:]) > 0:
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/opcode.go:2079]: >Do we slice the first byte off fullSigBytes because we know that we have at least one byte that corresponds to the hash type, so the remaining should be for the signature itself?

Yeah, before this PR, this fragment looked like: https://github.com/btcsuite/btcd/blob/81fbd9b67e5497963b089bd77288933e4f7657c6/txscript/opcode.go#L1966

At that point, `sigBytes` already has had the sighash byte removed from it. Open to ways to make this clearer or more explicit. It's one of the down stream effects of needing to explicitly deal with 64 vs 65 byte signatures unfortunately. 

> Interesting how a nil push (no sig, no hash type) with ScriptVerifyNullFail doesn't halt execution but this case does.

The null fail semantics are a bit tricky implementation wise. IMO the most straight forward statement is that: if a checksig is failing due to validation, then the signature must be an empty byte array. 

It comes up more in multi-sig settings where old scripts didn't have tapscript paths, so may be trying multiple multi-sig thresholds to see which one passes. W/o nullfail, some one could mutate sigs that weren't needed which then modifies the wtxid.

File: txscript/opcode.go [hunk 16]
```diff
@@ -2094,6 +2094,84 @@ func opcodeCheckSigVerify(op *opcode, data []byte, vm *Engine) error {
...
+	}
+	sigBytes, err := vm.dstack.PopByteArray()
+	if err != nil {
+		return err
+	}
+
+	// If the signature is empty, then we'll just push the value N back
+	// onto the stack and continue from here.
+	//
+	// TODO(roasbeef): unknown key check before this?
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/opcode.go:2149]: Hmm, this seems a bit insufficiently specified... 

If the signature is a nil vector, then it'll fail during validation above, meaning execution halts and an empty vector isn't pushed onto the stack. Therefor we need to check this here so we just thread through the accumulator without any further actions. Alternatively, we can defer parsing the signature until the very end. 

Will make sure to confirm this behavior before moving from the draft state for this PR.

**Comment by Roasbeef** [comment]
Just pushed a revamped version which addresses all existing review comments, and also commits the `bitcoind` taproot test vectors. With the latest set of commits, we pass all 2760 test cases (both success and reject sub-cases for each test case). I've now removed the "draft" label from this PR, so we can consider it to be more final-ish! 

I have some smaller TODO's to clean up still, like better error messages, but I'm pretty confident/happy in the current diff. Still entertaining the option of splitting this up into smaller PRs as with the latest diff (ignore all the files committed as those are just test vectors) the PR size has swelled a bit.

**Comment by 3nprob** [comment]
The little I've been looking through, this looks great!

Splitting up into smaller PRs would help a lot, if you do have the inclination ^^

File: txscript/sign.go [hunk 17]
```diff
@@ -114,6 +133,39 @@ func TaprootWitnessSignature(tx *wire.MsgTx, sigHashes *TxSigHashes, idx int,
 	return wire.TxWitness{sig}, nil
 }
 
+// RawTxInTapscriptSignature computes a raw schnorr signature for a signature
+// generated from a tapscript leaf. This differs from the
+// RawTxInTaprootSignature which is used to generate signatures for top-level
+// taproot key spends.
+//
+// TODO(roasbeef): actually add code-sep to interface? not really used
```
[the remark this answers is not in the document]
Comment (guggero) [reply txscript/sign.go:141]: Okay, I found the bug! The `codeSepPos` was wrong, see comment below.

File: txscript/sign.go [hunk 18]
```diff
@@ -114,6 +133,39 @@ func TaprootWitnessSignature(tx *wire.MsgTx, sigHashes *TxSigHashes, idx int,
...
+func RawTxInTapscriptSignature(tx *wire.MsgTx, sigHashes *TxSigHashes, idx int,
+	amt int64, pkScript []byte, tapLeaf TapLeaf, hashType SigHashType,
+	privKey *btcec.PrivateKey) ([]byte, error) {
+
+	// First, we'll start by compute the top-level taproot sighash.
+	tapLeafHash := tapLeaf.TapHash()
+	sigHash, err := calcTaprootSignatureHashRaw(
+		sigHashes, hashType, tx, idx,
+		NewCannedPrevOutputFetcher(pkScript, amt),
+		WithBaseTapscriptVersion(0, tapLeafHash[:]),
```
Comment (guggero) [inline txscript/sign.go:152]: This default value of `0` here is incorrect for the normal script spend. I needed to set it to `math.MaxUint32` in my case. See https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki#common-signature-message-extension, not sure where this would ever be something other than `0xffffffff`? I don't think I understand what "the opcode position of the last executed" means in this context. Is this in case we have multiple OP_CHECKSIG in the same script, we'd need to keep track of them and increment this value accordingly?

In any case, I think we should probably allow this to be set as a parameter here.

**Review by guggero — CHANGES_REQUESTED** [review]
Did a full review (third round) and was able to end-to-end test key spend, script (single leaf and multiple leaf) execution, including OP_CHECKSIG with schnorr signatures.

Found a few small issues but all in all I'd say this is very close to be ready! Given we have superb coverage of the _validation_ code thanks to the test cases from bitcoind I'd say this can be merged very soon.

Really great work on this, the code is super clean and easy to read/understand!

**Comment by guggero** [comment]
I fixed a few things I mentioned in the comments here: https://github.com/guggero/btcd/tree/1787-fixes feel free to incorporate directly into this PR.

File: txscript/sign.go (hunk 18 above)
Comment (Roasbeef) [reply txscript/sign.go:152]: Fixed this by having the `defaultTaprootSighashOptions` function use the max int as the default arg there, thanks for catching this!

File: txscript/sign.go [hunk 19]
```diff
@@ -114,6 +133,39 @@ func TaprootWitnessSignature(tx *wire.MsgTx, sigHashes *TxSigHashes, idx int,
...
+
+	// With the sighash constructed, we can sign it with the specified
+	// private key.
+	signature, err := schnorr.Sign(privKey, sigHash)
+	if err != nil {
+		return nil, err
+	}
+
+	// Finally, append the sighash type to the final sig.
+	return append(signature.Serialize(), byte(hashType)), nil
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/sign.go:166]: This is already there, maybe you were working off an older version of the branch?

[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/sign.go]: Fixed! The comment was right, but then this section of the diff didn't follow though. Planning on adding more tests here so we ensure we validate the witnesses/sigs we produce.

**Comment by Roasbeef** [comment]
@guggero thanks! Added those in, and also working on more tests to ensure we accept sigs/witnesses we generate.

**Review by guggero — APPROVED** [review]
End-to-end tested this latest iteration in `lnd` and all test cases were successful. Of course more unit tests are always great to get more coverage on the signing part.

Since the verification part is covered very well, I'm giving my LGTM so we can move forward with the depending PRs.

File: txscript/engine.go [hunk 20]
```diff
@@ -411,20 +427,22 @@ func checkMinimalDataPush(op *opcode, data []byte) error {
...
+	if vm.taprootCtx == nil && isOpcodeDisabled(op.value) {
 		str := fmt.Sprintf("attempt to execute disabled opcode %s", op.name)
 		return scriptError(ErrDisabledOpcode, str)
 	}
 
-	// Always-illegal opcodes are fail on program counter.
-	if isOpcodeAlwaysIllegal(op.value) {
+	// Always-illegal opcodes are fail on program counter, but only if
+	// we're not executing a taproot script.
+	if vm.taprootCtx == nil && isOpcodeAlwaysIllegal(op.value) {
```
Comment (wpaulino) [inline txscript/engine.go:439]: Why is there an exception for this with Taproot? The newly introduced success opcodes are all part of `isOpcodeDisabled`, so I can see why the exception exists there, but `OP_VERIF` and `OP_VERNOTIF` (illegal opcodes) are not part of that set.

**Review by wpaulino — COMMENTED** [review]
This is looking pretty good! Very happy to see such an extensive testing suite for these changes.

I'll be working on a small PoC leveraging the new APIs to produce and verify tapscript spends and report back any findings in my next review.

File: txscript/pkscript.go [hunk 21]
```diff
@@ -233,6 +240,8 @@ func computeNonWitnessPkScript(sigScript []byte) (PkScript, error) {
 
 // computeWitnessPkScript computes the script of an output by looking at the
 // spending input's witness.
+//
+// TODO(roasbeef): update for taproot based on sipa comment in neutrino repo
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/pkscript.go:244]: Gonna remove this TODO and make an issue for it in the tracker: https://github.com/btcsuite/btcd/issues/1817

File: blockchain/utxoviewpoint.go [hunk 22]
```diff
@@ -159,6 +159,19 @@ func (view *UtxoViewpoint) LookupEntry(outpoint wire.OutPoint) *UtxoEntry {
...
 
+// FetchPrevOutput fetches the previous output referenced by the passed
+// outpoint. This is identical to the LookupEntry method, but it returns a
+// wire.TxOut instead.
+//
+// NOTE: This is an implementation of the txscript.PrevOutputFetcher interface.
+func (view *UtxoViewpoint) FetchPrevOutput(op wire.OutPoint) *wire.TxOut {
+	prevOut := view.entries[op]
+	return &wire.TxOut{
+		Value:    prevOut.amount,
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply blockchain/utxoviewpoint.go:170]: True that usually at this point we've already verified that the UTXOs already exist in the valiration setting. This also comes up when computing sighashes, so a caller may make a mistake leading to either a panic at the call site or here. Modified to return a `nil` pointer so we avoid the panic _here_.

File: txscript/sigvalidate.go [hunk 23]
```diff
@@ -0,0 +1,480 @@
...
+	}
+
+	// Before we attempt to verify the signature, we'll need to first
+	// compute the sighash based on the input and tx information.
+	sigHash, err := calcTaprootSignatureHashRaw(
+		t.hashCache, t.hashType, t.tx, t.inputIndex, t.prevOuts,
+		opts...,
+	)
+	if err != nil {
+		// TODO(roasbeef): propagate the error here?
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/sigvalidate.go:376]: `calcTaprootSignatureHashRaw` only returns en error if:
  1.  An invalid sighash is used, bitcoind does the same here: https://github.com/bitcoin/bitcoin/blob/623745ca74cf3f54b474dac106f5802b7929503f/src/script/interpreter.cpp#L1716
  2. An index is passed in that's greater than the total number of actual inputs 
    * For this, earlier in their pipeline there's an assertion to fail early for this case. We have a return clause here as we expose this function to consumers of the library, which may pass in an invalid index. 

The other error paths there are just error returns from the `io.Writer` when creating the sig message, which in practice don't return an error.

File: txscript/taproot.go [hunk 24]
```diff
@@ -48,3 +90,113 @@ func VerifyTaprootKeySpend(witnessProgram []byte, rawSig []byte, tx *wire.MsgTx,
...
+	// The control block can't be larger than a proof for the largest
+	// possible tapscript merkle tree with 2^128 leaves.
+	case len(ctrlBlock) > ControlBlockMaxSize:
+		return nil, fmt.Errorf("invalid max block size")
+
+	// Ignoring the fixed sized portion, we expect the total number of
+	// remaining bytes to be a multiple of the node size, which is 32
+	// bytes.
+	case (len(ctrlBlock)-ControlBlockBaseSize)%ControlBlockNodeSize != 0:
+		return nil, fmt.Errorf("invalid max block size")
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/taproot.go:203]: Will resolve these in a final pass to insert proper errors.

File: txscript/engine.go (hunk 20 above)
Comment (Roasbeef) [reply txscript/engine.go:439]: Hmm, yeah you're right that this is just a case for those two particular op codes. Reflecting, I think I added this here to make it uniform with the check about ti ognire disabled op codes. Re the success op codes, execution is short circuited elsewhere now (before we even step through any op codes), so this can be safely removed (the guard around is opcode disabled).

**Comment by Roasbeef** [comment]
Responded to the latest round of comments! Also added a new `CalcTapscriptSignaturehash` that can be used to compute the sighash of tapscript spends. This'll be useful when doing things like validating that the remote party sent a valid HTLC signature on the LN context. 

On the `lnd` side, we've successfully done end-to-end tests of both keyspend and script spend paths (using [this](https://github.com/lightningnetwork/lnd/pull/6263) lnd PR and the btcwallet PR linked).

**Comment by Roasbeef** [comment]
Just pushed out two commits to add the taproot deployment, and also use the new rules after activation. This can actually go in another PR, but I think it's nice for people to be able to run this PR as a single unit, and also for us to run sync tests on mainnet+testnet to ensure there aren't any edge cases we missed along the way.

**Comment by Roasbeef** [comment]
Testnet sync was succesful, with the ST BIP 9 state machine switching over at the proper time. My mainnet sync should finish tomorrow-ish. Afterwards, I think this'll clear for a merge (just need to insert those final errors).

**Review by wpaulino — COMMENTED** [review]
Noticed a few unnecessary mutations happening in some of the new Taproot functions throughout developing my PoC. Other than that, I think this is ready to go barring any issues left w.r.t. mainnet node syncs.

**Comment by Roasbeef** [comment]
Responded to the latest comments, and pushed up a last commit to add replace directives since this PR modified the internal sub-modules. In the future, we should try to make this changes into an independent PR, and also attempt to see if tagging them within the branch works out when things are merged into mater as well.

**Comment by Roasbeef** [comment]
> Mainnet sync ✅

🚨 Just received word from Merge Control: this PR is a go!! 🚨
